Add Mellanox CI scripts to the project (#1729)
vmware-tanzu/antrea: ci/jenkins/mellanox/README.md

# Mellanox CI

> Mellanox CI is used to test hardware offload in the antrea project. It deploys Kubernetes with the antrea image and tests wither the OVS have hardware offload capabilities or not.

Reviewer: antrea -> Antrea

Reviewer: wither -> whether
OVS have -> OVS has

Reviewer: But is it to test whether OVS has offload capability, or test Antrea works well with OVS hardware offload?

> Mellanox CI is used to test hardware offload in the Antrea project. It deploys Kubernetes with the Antrea image and tests whether the OVS has hardware offload capabilities or not.

Reviewer 2: Could you wrap the lines at 80 characters in documentation files?

Author: Done for normal lines, but faced some issues with the table.
It seems it is not possible to split lines for markdown table cells, for now i left them as is, do you think we need to handle it?

Reviewer 3: Maybe you can use "<br />" to split a long line. Now comments column can be hardly seen...
https://stackoverflow.com/questions/11700487/newline-in-markdown-table

## Workflow

The CI uses three scripts to run the tests:

> * start_ci.sh: Build and deploy the Kubernetes test environment including Antrea.

Reviewer 2: I took a quick look at the scripts and it seems that the K8s cluster used for testing is a single-Node cluster and that the version used is the latest stable release (as per `https://storage.googleapis.com/kubernetes-release/release/stable.txt`). I think this is worth documenting here.

Author: Added a small description of the environment at the start of the section and put the details there.
Please have a look.

> * start_ci.sh: Build and deploy the kubernetes test environment including the antrea.

Reviewer: kubernetes -> Kubernetes

Reviewer: the antrea -> Antrea

* test.sh: Test the hardware offload.
* stop_ci.sh: Stop and clean the testing environment.

The CI can be triggered by commenting `/test-hw-offload` on an opened pull request.

## Configurations

> The scripts use Bash environment variables to configure its behavior. The following table shows what variables can be configured:

Reviewer 2: s/its behavior/their behavior
(because scripts)

Author: Done

> The scripts use bash environment variables to configure its behavior. The following table shows what variables can be configured:

Reviewer: Probably: bash -> Bash

|  Variable |  DEFAULT VALUE |  Comments |
|  ------ |  ------ |  ------ |

> |  RECLONE | true | whether or not to reclone projects in case of single workspace |

Reviewer: I saw other lines start with upper case, so: whether -> Whether.

|  WORKSPACE | /tmp/k8s | The directory that willcontain all the project components |
|  LOGDIR | $WORKSPACE/logs | The directory where the logs should be put |
|  ARTIFACTS | $WORKSPACE/artifacts | The directory where configuration files should be put |

> |  TIMEOUT | 300 | Timeout time for pods status |

Reviewer: pods -> Pods

|  POLL_INTERVAL | 10 | The interval to wait between each check for pods status change |

> |  KUBERNETES_VERSION | latest_stable | The kubernetes version (or branch) to build |

Reviewer: kubernetes -> Kubernetes

> |  HARBOR_REGESTRY | harbor.mellanox.com | The docker registry to use to pull images from |

Reviewer: Maybe remove "to use"?

> |  HARBOR_REGESTRY | harbor.mellanox.com | The docker registry to pull images from |

Reviewer 2: it's spelled "registry"

Author: Done. Changed in the table and in the scripts.

|  HARBOR_PROJECT | cloud-orchestration | The docker registry project to use to pull the images from |

> |  MULTUS_CNI_REPO | <https://github.com/intel/multus-cni> | multus cni repo URL |

Reviewer: multus -> Multus

Reviewer: cni -> CNI

|  MULTUS_CNI_BRANCH | master | multus cni branch to build |

> |  MULTUS_CNI_PR || multus cni pr to pull, if this is used the MULTUS_CNI_BRANCH is ignored |

Reviewer: pr -> PR or pull request

|  MULTUS_CNI_HARBOR_IMAGE | ${HARBOR_REGESTRY}/${HARBOR_PROJECT}/multus | The multus image to pull in case the project was not built |
|  PLUGINS_REPO | <https://github.com/containernetworking/plugins.git> | containernetworking repo URL |
|  PLUGINS_BRANCH | master | containernetworking branch to build |
|  PLUGINS_BRANCH_PR || containernetworking cni pr to pull, if this is used the PLUGINS_BRANCH is ignored |
|  SRIOV_NETWORK_DEVICE_PLUGIN_REPO | <https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin> | SRIOV network device plugin repo to use |
|  SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH | master | SRIOV network device plugin branch to build |
|  SRIOV_NETWORK_DEVICE_PLUGIN_PR || SRIOV network device plugin pull request to pull, adding this will ignore SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH |
|  SRIOV_NETWORK_DEVICE_PLUGIN_HARBOR_IMAGE | ${HARBOR_REGESTRY}/${HARBOR_PROJECT}/sriov-device-plugin | The sriov-network-device-plugin image to pull in case the project was not built |
|  GOPATH | ${WORKSPACE} ||
|  PATH | /usr/local/go/bin/:$GOPATH/src/k8s.io/kubernetes/third_party/etcd:$PATH ||

> |  CNI_BIN_DIR | /opt/cni/bin/ | this is used to configure Kubernetes local_cluser_up.sh CNI_BIN_DIR |

Reviewer: this -> This

|  CNI_CONF_DIR | /etc/cni/net.d/ | this is used to configure Kubernetes local_cluser_up.sh CNI_CONF_DIR |
|  API_HOST | $(hostname) | The node name to use |
|  API_HOST_IP | $(hostname -I | awk '{print $1}') | The API server IP |

> |  POD_CIDER | 192.168.0.0/16 | The pods network subnet |

Reviewer: pods -> Pods

> |  POD_CIDER | 192.168.0.0/16 | The pods network subnet |

Reviewer: CIDER -> CIDR

> |  SERVICE_CIDER | 172.0.0.0/16 | The service network subnet |

Reviewer: service -> Service

|  KUBECONFIG | /etc/kubernetes/admin.conf | The KUBECONFIG file to use |
|  SRIOV_INTERFACE | auto_detect | The Mellanox interface to use to create the VFs |

> |  NETWORK | "192.168.$N" | this is used to setup the macvlan network range, N is randomly generated |

Reviewer: this -> This
macvlan -> MACVLAN

> |  ANTREA_CNI_REPO | <https://github.com/vmware-tanzu/antrea.git> | antrea project repo to use |

Reviewer: antrea -> Antrea

|  ANTREA_CNI_BRANCH | master | Antrea project branch to use |
|  ANTREA_CNI_PR | | Antrea project pull request number to pull, adding this will ignore ANTREA_CNI_BRANCH |

> |  ANTREA_CNI_HARBOR_IMAGE | ${HARBOR_REGESTRY}/${HARBOR_PROJECT}/antrea | The Antrea image to pull in case the project was not built |

Reviewer 2: just want to confirm that `ANTREA_CNI_HARBOR_IMAGE` is not used when testing a pull request, since then the image would be built?

Author: That is correct, the images is not used in case of pull requests. The image is built nightly from the Antrea master branch and used in a daily CI. For pull requests the ci will try to build the image from the fetched pull request, and uses the built image. see <https://github.com/vmware-tanzu/antrea/pull/1729/files?file-filters%5B%5D=.sh#diff-12b01860bd03aec02f37d67c6cec364f88591242e1ad900fe68e915ba3a41ec1R73-R87>

# Mellanox CI

Mellanox CI is used to test hardware offload in the Antrea project.
 It deploys Kubernetes with the Antrea image and tests whether the

>  OVS has hardware offload capabilities or not.

Reviewer 2: I don't understand the extra white spaces at the beginning of lines throughout this file?

Author: sorry, my bad. removed them.

## Workflow

The CI tests run on a single node cluster with a ConnectX5 NIC card,
 the latest stable Kubernetes binaries are built each time the scripts

>  run. In addition to the Kubernetes and the Antrea, the scripts deploy

Reviewer 2: s/In addition to the Kubernetes and the Antrea/In addition to Kubernetes and Antrea

Author: Done

 Multus CNI, SRIOV device plugin, and container networking plugins. The
 versions for the used components are summarized as follow:

* Kubernetes: Latest stable.
* Antrea: Image built from pull request.
* Multus: A nightly pulled latest stable Multus image.

> * SRIOV device plugin: A nightly built master image.

Reviewer 2: ```suggestion
* SRIOV device plugin: A nightly built image from the main branch.
```

Author: My bad, forgot about the naming convention. Done

> * container networking plugins: latest master binaries.

Reviewer 2: ```suggestion
* container networking plugins: Latest binaries built from the main branch.
```

Author: Done

|  PLUGINS_BRANCH | master | containernetworking branch to build |
|  PLUGINS_BRANCH_PR || containernetworking pull request number to pull, if this is used the PLUGINS_BRANCH is ignored |
|  SRIOV_NETWORK_DEVICE_PLUGIN_REPO | <https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin> | SRIOV network device plugin repository to use |
|  SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH | master | SRIOV network device plugin branch to build |
|  SRIOV_NETWORK_DEVICE_PLUGIN_PR || SRIOV network device plugin pull request number to pull, adding this will ignore SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH |
|  SRIOV_NETWORK_DEVICE_PLUGIN_HARBOR_IMAGE | ${HARBOR_REGISTRY}/${HARBOR_PROJECT}/sriov-device-plugin | The SRIOV network device plugin image to pull in case the project was not built |
|  GOPATH | ${WORKSPACE} ||
|  PATH | /usr/local/go/bin/:$GOPATH/src/k8s.io/kubernetes/third_party/etcd:$PATH ||
|  CNI_BIN_DIR | /opt/cni/bin/ | This is used to configure Kubernetes local_cluser_up.sh CNI_BIN_DIR |
|  CNI_CONF_DIR | /etc/cni/net.d/ | This is used to configure Kubernetes local_cluser_up.sh CNI_CONF_DIR |
|  API_HOST | $(hostname) | The Node name to use |

> |  API_HOST_IP | $(hostname -I | awk '{print $1}') | The API server IP |

Reviewer 3: ```suggestion
|  API_HOST_IP | $(hostname -I \| awk '{print $1}') | The API server IP |
```
| should be escaped.

Author: Done
